engine/consensus/dkg/reactor_engine.go

```diff
@@ -161,33 +229,56 @@ func (e *ReactorEngine) EpochSetupPhaseStarted(currentEpochCounter uint64, first
-// key vectors. If the keys don't match a log statement will be invoked. In the happy case the locally
-// produced key should match, if the keys do not match the node will have a invalid random beacon key.
-func (e *ReactorEngine) EpochCommittedPhaseStarted(currentEpochCounter uint64, first *flow.Header) {
+// handleEpochCommittedPhaseStarted checks that the local DKG completed and
+// that our locally computed key share is consistent with the canonical key vector.
+//
+// If our key is inconsistent, we log a warning message.
+// CAUTION: we will still sign using the inconsistent key, to be fixed with Crypto
+// v2 by falling back to staking signatures in this case.
+//
+func (e *ReactorEngine) handleEpochCommittedPhaseStarted(currentEpochCounter uint64) {
+
+	// the epoch counter the DKG just completed is for
+	dkgEpochCounter := currentEpochCounter + 1
+
+	// TODO - first check whether we've already stored the dkg end state, since
+	// we need to handle multiple calls to this method
+
+	// TODO execute in transaction scope of protocol event emitter, need to use
+	// lower-level storage access here (since State doesn't allow tx injection)
 	nextDKG, err := e.State.Final().Epochs().Next().DKG()
 	if err != nil {
-		e.log.Err(err).Msg("checking DKG key consistency: could not retrieve next DKG info")
+		// CAUTION: this should never happen, indicates a storage failure or corruption
+		e.log.Fatal().Err(err).Msg("checking DKG key consistency: could not retrieve next DKG info")
 		return
 	}
 
-	dkgPrivInfo, err := e.keyStorage.RetrieveMyBeaconPrivateKey(currentEpochCounter + 1)
-	if err != nil {
-		e.log.Err(err).Msg("checking DKG key consistency: could not retrieve DKG private info for next epoch")
+	myBeaconPrivKey, err := e.dkgState.RetrieveMyBeaconPrivateKey(currentEpochCounter + 1)
+	if errors.Is(err, storage.ErrNotFound) {
+		e.log.Warn().Msg("checking DKG key consistency: no key found")
+		err := e.dkgState.SetDKGEndState(dkgEpochCounter, flow.DKGEndStateNoKey)
+		if err != nil {
+			e.log.Fatal().Err(err).Msg("failed to set dkg end state")
+		}
+	} else if err != nil {
+		e.log.Fatal().Err(err).Msg("checking DKG key consistency: could not retrieve DKG private key for next epoch")
 		return
 	}
 
 	nextDKGPubKey, err := nextDKG.KeyShare(e.me.NodeID())
 	if err != nil {
-		e.log.Err(err).Msg("checking DKG key consistency: could not retrieve DKG public key for next epoch")
+		e.log.Fatal().Err(err).Msg("checking DKG key consistency: could not retrieve my DKG public key for next epoch")
 		return
 	}
 
-	localPubKey := dkgPrivInfo.PublicKey()
+	localPubKey := myBeaconPrivKey.PublicKey()
 
 	if !nextDKGPubKey.Equals(localPubKey) {
 		e.log.Warn().Msg("checking DKG key consistency: locally computed dkg public key does not match dkg public key for next epoch")
+		err := e.dkgState.SetDKGEndState(dkgEpochCounter, flow.DKGEndStateInconsistentKey)
+		if err != nil {
+			e.log.Fatal().Err(err).Msg("failed to set dkg end state")
```

Comment:
Better to add current epoch in the error message, same for other errors in this function

```suggestion
			e.log.Fatal().Uint64("cur_epoch", currentEpochCounter).Err(err).Msg("failed to set dkg end state")
```

Reply:
https://github.com/onflow/flow-go/pull/1721/commits/b5e69b4507024026c63471c93ecd41229b4f5365